library/core/src/marker.rs

```diff
@@ -724,9 +724,9 @@ impl<T: ?Sized> !Sync for *mut T {}
 ///
 /// The exact interaction of `PhantomData` with drop check **may change in the future**.
 ///
-/// Currently, adding a field of type `PhantomData<T>` indicates that your type *owns* data of type
-/// `T` in very rare circumstances. This in turn has effects on the Rust compiler's [drop check]
-/// analysis. For the exact rules, see the [drop check] documentation.
+/// Currently, adding a field of type `PhantomData<T>` indicates the type *owns* data of type `T`.
+/// In very rare circumstances, this has effects on the Rust compiler's [drop check] analysis.
```

Comment:
It still mentions the rare circumstances, so I don't think this fixes the issue. I don't like that sentence either. It should either be dropped, or explicitly mention those rare circumstances ("using the unstable may_dangle feature").

Reply:
Hmm. My understanding is that the "rare circumstances" ambiguity comes about because the teams are hedging so they are allowed to change it.

My draft seems more acceptable to me than the status quo because it's unambiguous as to _whether and when_ you _should_ use PhantomData, it just is sly about whether it has any given effect.

Reply:
cc https://github.com/rust-lang/rust/pull/103413#discussion_r1616312771

"Currently, adding a field of type `PhantomData<T>` indicates that your type *owns* data of type `T` in very rare circumstances" is intentional. Adding `PhantomData<T>` to your type only impacts whether your type 'owns' `T` in rare circumstances: if your type already has another field which requires drop:

```rust
struct DoesNotOwnT<T> {
   field: *const T,
   phantom_data: PhantomData<T>,
}
```

```rust
struct DoesOwnT<T> {
   field: *const T,
   other_field: String,
   phantom_data: PhantomData<T>,
}
```

Reply:
I've opened an RFC to fix this behavior https://github.com/rust-lang/rfcs/pull/3417, but I currently don't have the capacity to push this over the finish line

Reply:
oic, thank you for explaining!

hmm... the definition provided here still feels _abstruse_, but I understand why you phrased it that way now.

Reply:
yeah 🤔 I think for this to make sense you'd have to actually go into the difference between `needs_drop` and `dropck_outlives` (`PhantomData` is only considered for `dropck_outlives`, as there shouldn't be any types for which `needs_drop` and `Copy` both hold)

Ideally the way we make this more understandable is by just merging https://github.com/rust-lang/rfcs/pull/3417, though I guess we can also go more in-depth in the comment if you think that's more achievable.

Reply:
Hmm, I currently do not have an idea of how to proceed!